### PR TITLE
manifests: run gateways as non-root user

### LINKED ISSUE #23072: Unable to run istio-ingressgateway as nonroot user (1337)
**Bug description**
The istio-ingressgateway enters CrashLoopBackOff when trying to run it as nonroot user due to file permission issues with /etc/istio/proxy/envoy-rev0.json. The deployment has been patched with:
```
 spec:
  template:
    spec:
      securityContext:
        runAsUser: 1337
        runAsGroup: 1337
        fsGroup: 1337
```
I have seen some issues/threads on this that claimed that it can run as user istio-proxy, however it does not seem to work.

```
2020-04-20T10:12:43.896950Z info parsed scheme: ""

2020-04-20T10:12:43.896965Z info scheme "" not registered, fallback to default scheme

2020-04-20T10:12:43.896983Z info ccResolverWrapper: sending update to cc: {[{istio-pilot.istio-system.svc:15012  <nil> 0 <nil>}] <nil> <nil>}

2020-04-20T10:12:43.896994Z info ClientConn switching balancer to "pick_first"

2020-04-20T10:12:43.897532Z info pickfirstBalancer: HandleSubConnStateChange: 0xc0006f6ba0, {CONNECTING <nil>}

2020-04-20T10:12:43.904553Z info pickfirstBalancer: HandleSubConnStateChange: 0xc0006f6ba0, {READY <nil>}

2020-04-20T10:12:43.931696Z info Starting gateway SDS

2020-04-20T10:12:44.033513Z error sds Failed to listen on unix socket "/etc/istio/proxy/SDS": listen unix /etc/istio/proxy/SDS: bind: permission denied

2020-04-20T10:12:44.033595Z error sds Failed to set up UDS path: listen unix /etc/istio/proxy/SDS: bind: permission denied

2020-04-20T10:12:44.033629Z info sds SDS gRPC server for workload UDS starts, listening on "/etc/istio/proxy/SDS" 


2020-04-20T10:12:44.033688Z info sds Start SDS grpc server

2020-04-20T10:12:44.033791Z error sds Failed to listen on unix socket "/etc/istio/proxy/SDS": listen unix /etc/istio/proxy/SDS: bind: permission denied

2020-04-20T10:12:44.033804Z error sds SDS grpc server for workload proxies failed to set up UDS: listen unix /etc/istio/proxy/SDS: bind: permission denied

2020-04-20T10:12:44.033851Z info sds SDS gRPC server for ingress gateway controller starts, listening on "/var/run/ingress_gateway/sds" 


2020-04-20T10:12:44.033894Z info PilotSAN []string{"istio-pilot.istio-system.svc"}

2020-04-20T10:12:44.033912Z info Starting proxy agent

2020-04-20T10:12:44.033994Z info Opening status port 15020


2020-04-20T10:12:44.033987Z info sds Start SDS grpc server for ingress gateway proxy

2020-04-20T10:12:44.034021Z info Received new config, creating new Envoy epoch 0

2020-04-20T10:12:44.034043Z info Epoch 0 starting

**2020-04-20T10:12:44.034083Z error Failed to generate bootstrap config: open /etc/istio/proxy/envoy-rev0.json: permission **
```
In https://github.com/istio/istio/blob/release-1.5.1-patch/pilot/docker/Dockerfile.proxyv2 there is chown command executed as follows:
```
RUN chown -R istio-proxy /var/lib/istio
```

Is there any reason for not doing the same ownership change for /etc/istio/proxy/ ?
With a custom image I was able to get istio-ingressgateway working when changing ownership of that directory.

Am I missing something here or this behavior is not intended? I would rather not divert from the upstream but this is a crucial security feature for me and many other users I believe to be able to run ingressgateway as nonroot. Seems like an easy fix unless there are reasons for that directory not being owned by istio-proxy.

**Expected behavior**
The istio-ingressgateway runs fine as nonroot user.

**Steps to reproduce the bug**
```
istioctl manifest apply --set values.gateways.istio-ingressgateway.nodeSelector.role=worker
kubectl patch deployment istio-ingressgateway --type merge -p "$(cat istio.yaml)" -n istio-system
```
istio.yaml
```
 spec:
  template:
    spec:
      securityContext:
        runAsUser: 1337
        runAsGroup: 1337
        fsGroup: 1337
```

**Version**
Istio:
```
client version: 1.5.1
control plane version: 1.5.1
data plane version: 1.5.1 (2 proxies)
```

In addition, tried on 1.3.0, 1.4.7 and 1.5.0 with mostly the same results (versions pre 1.5.0 didn't complain about other things like permissions to sockets)

Kubernetes:
```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.1", GitCommit:"7879fc12a63337efff607952a323df90cdc7a335", GitTreeState:"clean", BuildDate:"2020-04-08T17:38:50Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.8-docker-1", GitCommit:"8100f4dfe656d4a4e5573fe86375a5324771ec6b", GitTreeState:"clean", BuildDate:"2019-10-18T17:13:51Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
```
**How was Istio installed?**
```
istioctl manifest apply --set values.gateways.istio-ingressgateway.nodeSelector.role=worker
```
**Environment where bug was observed (cloud vendor, OS, etc)**
```
Docker Enterprise 3.2.6
Centos 7.7.1908
```

### PR DESCRIPTION
**Pull request by dgn**
Fixes #23072, #20853

This is an attempt to revive the efforts in https://github.com/istio/istio/pull/20893

### CODE REVIEW

File: manifests/charts/gateways/istio-ingress/templates/deployment.yaml [hunk 1]
```diff
@@ -46,6 +46,11 @@ spec:
 {{ toYaml $gateway.podAnnotations | indent 8 }}
 {{ end }}
     spec:
+      securityContext:
```
Comment (costinm) [inline manifests/charts/gateways/istio-ingress/templates/deployment.yaml:49]: Please add a setting and wrap it.

**Comment by costinm** [comment]
This will require changes in pilot as well - including some metadata so pilot knows that a particular gateway is not running as root. And off by default at first (as all other new risky things), with 1-2 releases to turn it on by default. 

I love running as non-root - but we still need to follow the rules.

**Comment by howardjohn** [comment]
Why does pilot need to know if its root or not? It already listens on 8080 and 8443 - pilot handles targetPort. That change was made a while ago

**Comment by dgn** [comment]
> This will require changes in pilot as well - including some metadata so pilot knows that a particular gateway is not running as root. And off by default at first (as all other new risky things), with 1-2 releases to turn it on by default.
> 
> I love running as non-root - but we still need to follow the rules.

Agreed, I can add a feature flag - it might break customers with custom gateway ports <1024 that don't set a targetPort >1023.

I'm not sure why this would require pilot changes though- as John already mentioned, the relevant changes in Pilot to respect the targetPort have been merged long ago.

File: manifests/charts/gateways/istio-egress/templates/deployment.yaml [hunk 2]
```diff
@@ -186,6 +196,8 @@ spec:
           - name: ISTIO_META_CLUSTER_ID
             value: "{{ $.Values.global.multiCluster.clusterName | default `Kubernetes` }}"
           volumeMounts:
+          - name: istioproxyconfig
```
Comment (howardjohn) [inline manifests/charts/gateways/istio-egress/templates/deployment.yaml:199]: nit: make this `istio-envoy` to align with injection template. If there is a compelling reason not to it should maybe be istio-proxy-config to at least align with the formatting?

Comment (dgn) [reply manifests/charts/gateways/istio-egress/templates/deployment.yaml]: Done!

**Review by howardjohn — APPROVED** [review]

File: manifests/charts/gateways/istio-egress/templates/deployment.yaml [hunk 3]
```diff
@@ -46,6 +46,13 @@ spec:
 {{ toYaml $gateway.podAnnotations | indent 8 }}
 {{ end }}
     spec:
+{{- if not $gateway.runAsRoot }}
+      securityContext:
```
Comment (rlenglet) [inline manifests/charts/gateways/istio-egress/templates/deployment.yaml:50]: Please consider also setting the other fields in this `securityContext` to tighten the security, [like we do for sidecar proxies](https://github.com/kubernetes/enhancements/pull/1003#issuecomment-619087019):
```
        allowPrivilegeEscalation: false
        capabilities:
          drop:
          - ALL
        privileged: false
        readOnlyRootFilesystem: true
        runAsGroup: 1337
        fsGroup: 1337
        runAsNonRoot: true
        runAsUser: 1337
```

**Review by rlenglet — COMMENTED** [review]
LGTM

**Comment by Atreus-Technologies** [comment]
Just a brief note on whether this should be on by default - I just upgraded to 1.5.4 from 1.4.3 and discovered that all of my 12 gateways across 7 namespaces now require an additional RBAC config to give them access to privileged PSPs, which was a bit surprising. 

I'm still working through quite how they were running without root privileges before, as it strikes me as a bit odd.

Can I contribute some tests somewhere to ensure that components do not suddenly expect to run as root in future when they've been previously happy as non-root? This is clearly an important area, given that the Istio CNI is designed to allow the control plane to run as non-root; it would be good to ensure the use case is better tested in future.

Also wondering which IstioOperator field would be changed to set the gateways to run as non-root? And which version would that setting be operable from?

**Comment by rlenglet** [comment]
@Atreus-Technologies can you please create a GitHub issue to clarify what your problem is? There seems to be something odd with your setup. Gateways were running as root, and now they don't anymore.

**Comment by howardjohn** [comment]
Isn't the default setting here runAsRoot=true though (not matching PR
title)?

**Comment by rlenglet** [comment]
> Isn't the default setting here runAsRoot=true though (not matching PR
title)?

Indeed, cf. https://github.com/istio/istio/pull/23174/files#diff-d09cefdb8b19cd0d48e03d3a9f216335R108!
But then the behavior is not changed at all when that is set to `true`.  🤨

**Comment by howardjohn** [comment]
maybe coredump is enabled? Only thing I can think of

**Comment by Atreus-Technologies** [comment]
OK, so a few errors to confess on my part here, now that I'm looking at this, post-upgrade, and in the cold light of day.

1. It looks like I was previously manually deploying gateway workloads, so the version must have been exactly the same between both deployments as they weren't controlled by the Istio operator/istioctl - issues can't have been caused by the upgrade as I'd initially supposed.
2. I must have previously had some sort of privileged psp sitting around that I'd previously created which enabled the gateway - there's some evidence for that in the namespaces I checked, so I'm guessing this is also my mistake.

Basically, looks like the issues I was encountering were due to my own misconfiguration - so erm, carry on as you were folks! :blush:

**Comment by Atreus-Technologies** [comment]
Oh, but just as a side note, I can't see `securityContext`s anywhere in the default deployments, so I'm guessing that this still isn't the default setting... If someone can let me know where to set the `runasroot=false` that @rlenglet mentioned, I'd be most grateful (assuming this setting is usable in the 1.5.4 release).

**Comment by dgn** [comment]
> Oh, but just as a side note, I can't see `securityContext`s anywhere in the default deployments, so I'm guessing that this still isn't the default setting... If someone can let me know where to set the `runasroot=false` that @rlenglet mentioned, I'd be most grateful (assuming this setting is usable in the 1.5.4 release).

Hey, sorry but this is only available in 1.6+ versions, becoming the default in 1.7
